// Package errors contains errors for the domain "neb".
//
// This file is automatically generated by errawr-gen. Do not modify it.
package errors

import (
	errawrgo "github.com/puppetlabs/errawr-go"
	impl "github.com/puppetlabs/errawr-go/impl"
)

// Error is the type of all errors generated by this package.
type Error interface {
	errawrgo.Error
}

// External contains methods that can be used externally to help consume errors from this package.
type External struct{}

// API is a singleton instance of the External type.
var API External

// Domain is the general domain in which all errors in this package belong.
var Domain = &impl.ErrorDomain{
	Key:   "neb",
	Title: "Nebula",
}

// CommandSection defines a section of errors with the following scope:
// Command execution errors
var CommandSection = &impl.ErrorSection{
	Key:   "command",
	Title: "Command execution errors",
}

// CommandNoCommandToExecuteErrorCode is the code for an instance of "no_command_to_execute_error".
const CommandNoCommandToExecuteErrorCode = "neb_command_no_command_to_execute_error"

// IsCommandNoCommandToExecuteError tests whether a given error is an instance of "no_command_to_execute_error".
func IsCommandNoCommandToExecuteError(err errawrgo.Error) bool {
	return err != nil && err.Is(CommandNoCommandToExecuteErrorCode)
}

// IsCommandNoCommandToExecuteError tests whether a given error is an instance of "no_command_to_execute_error".
func (External) IsCommandNoCommandToExecuteError(err errawrgo.Error) bool {
	return IsCommandNoCommandToExecuteError(err)
}

// CommandNoCommandToExecuteErrorBuilder is a builder for "no_command_to_execute_error" errors.
type CommandNoCommandToExecuteErrorBuilder struct {
	arguments impl.ErrorArguments
}

// Build creates the error for the code "no_command_to_execute_error" from this builder.
func (b *CommandNoCommandToExecuteErrorBuilder) Build() Error {
	description := &impl.ErrorDescription{
		Friendly:  "A valid command was not found to execute",
		Technical: "A valid command was not found to execute",
	}

	return &impl.Error{
		ErrorArguments:   b.arguments,
		ErrorCode:        "no_command_to_execute_error",
		ErrorDescription: description,
		ErrorDomain:      Domain,
		ErrorMetadata:    &impl.ErrorMetadata{},
		ErrorSection:     CommandSection,
		ErrorSensitivity: errawrgo.ErrorSensitivityNone,
		ErrorTitle:       "No command available to execute",
		Version:          1,
	}
}

// NewCommandNoCommandToExecuteErrorBuilder creates a new error builder for the code "no_command_to_execute_error".
func NewCommandNoCommandToExecuteErrorBuilder() *CommandNoCommandToExecuteErrorBuilder {
	return &CommandNoCommandToExecuteErrorBuilder{arguments: impl.ErrorArguments{}}
}

// NewCommandNoCommandToExecuteError creates a new error with the code "no_command_to_execute_error".
func NewCommandNoCommandToExecuteError() Error {
	return NewCommandNoCommandToExecuteErrorBuilder().Build()
}

// CommandUnknownCommandExecutionErrorCode is the code for an instance of "unknown_command_execution_error".
const CommandUnknownCommandExecutionErrorCode = "neb_command_unknown_command_execution_error"

// IsCommandUnknownCommandExecutionError tests whether a given error is an instance of "unknown_command_execution_error".
func IsCommandUnknownCommandExecutionError(err errawrgo.Error) bool {
	return err != nil && err.Is(CommandUnknownCommandExecutionErrorCode)
}

// IsCommandUnknownCommandExecutionError tests whether a given error is an instance of "unknown_command_execution_error".
func (External) IsCommandUnknownCommandExecutionError(err errawrgo.Error) bool {
	return IsCommandUnknownCommandExecutionError(err)
}

// CommandUnknownCommandExecutionErrorBuilder is a builder for "unknown_command_execution_error" errors.
type CommandUnknownCommandExecutionErrorBuilder struct {
	arguments impl.ErrorArguments
}

// Build creates the error for the code "unknown_command_execution_error" from this builder.
func (b *CommandUnknownCommandExecutionErrorBuilder) Build() Error {
	description := &impl.ErrorDescription{
		Friendly:  "something unexpected happened with command exection",
		Technical: "something unexpected happened with command exection",
	}

	return &impl.Error{
		ErrorArguments:   b.arguments,
		ErrorCode:        "unknown_command_execution_error",
		ErrorDescription: description,
		ErrorDomain:      Domain,
		ErrorMetadata:    &impl.ErrorMetadata{},
		ErrorSection:     CommandSection,
		ErrorSensitivity: errawrgo.ErrorSensitivityNone,
		ErrorTitle:       "Unknown command execution error",
		Version:          1,
	}
}

// NewCommandUnknownCommandExecutionErrorBuilder creates a new error builder for the code "unknown_command_execution_error".
func NewCommandUnknownCommandExecutionErrorBuilder(errorString string, commandString string) *CommandUnknownCommandExecutionErrorBuilder {
	return &CommandUnknownCommandExecutionErrorBuilder{arguments: impl.ErrorArguments{
		"command_string": impl.NewErrorArgument(commandString, "Command being executed"),
		"error_string":   impl.NewErrorArgument(errorString, "Error being thrown"),
	}}
}

// NewCommandUnknownCommandExecutionError creates a new error with the code "unknown_command_execution_error".
func NewCommandUnknownCommandExecutionError(errorString string, commandString string) Error {
	return NewCommandUnknownCommandExecutionErrorBuilder(errorString, commandString).Build()
}

// WorkflowSection defines a section of errors with the following scope:
// Workflow errors
var WorkflowSection = &impl.ErrorSection{
	Key:   "workflow",
	Title: "Workflow errors",
}

// WorkflowUnknownRuntimeErrorCode is the code for an instance of "unknown_runtime_error".
const WorkflowUnknownRuntimeErrorCode = "neb_workflow_unknown_runtime_error"

// IsWorkflowUnknownRuntimeError tests whether a given error is an instance of "unknown_runtime_error".
func IsWorkflowUnknownRuntimeError(err errawrgo.Error) bool {
	return err != nil && err.Is(WorkflowUnknownRuntimeErrorCode)
}

// IsWorkflowUnknownRuntimeError tests whether a given error is an instance of "unknown_runtime_error".
func (External) IsWorkflowUnknownRuntimeError(err errawrgo.Error) bool {
	return IsWorkflowUnknownRuntimeError(err)
}

// WorkflowUnknownRuntimeErrorBuilder is a builder for "unknown_runtime_error" errors.
type WorkflowUnknownRuntimeErrorBuilder struct {
	arguments impl.ErrorArguments
}

// Build creates the error for the code "unknown_runtime_error" from this builder.
func (b *WorkflowUnknownRuntimeErrorBuilder) Build() Error {
	description := &impl.ErrorDescription{
		Friendly:  "an unknown error occurred",
		Technical: "an unknown error occurred",
	}

	return &impl.Error{
		ErrorArguments:   b.arguments,
		ErrorCode:        "unknown_runtime_error",
		ErrorDescription: description,
		ErrorDomain:      Domain,
		ErrorMetadata:    &impl.ErrorMetadata{},
		ErrorSection:     WorkflowSection,
		ErrorSensitivity: errawrgo.ErrorSensitivityNone,
		ErrorTitle:       "Unknown runtime error",
		Version:          1,
	}
}

// NewWorkflowUnknownRuntimeErrorBuilder creates a new error builder for the code "unknown_runtime_error".
func NewWorkflowUnknownRuntimeErrorBuilder() *WorkflowUnknownRuntimeErrorBuilder {
	return &WorkflowUnknownRuntimeErrorBuilder{arguments: impl.ErrorArguments{}}
}

// NewWorkflowUnknownRuntimeError creates a new error with the code "unknown_runtime_error".
func NewWorkflowUnknownRuntimeError() Error {
	return NewWorkflowUnknownRuntimeErrorBuilder().Build()
}
